src/main/java/io/aiven/kafka/connect/common/output/SinkSchemaBuilder.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.connect.common.output;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.header.Header;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import io.aiven.kafka.connect.common.config.OutputField;
+
+import io.confluent.connect.avro.AvroData;
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/* Build schema for output fields based on SinkRecord values
+ * For multiple output fields schema looks like this:
+ * {
+ *    "type" "record", "fields": [
+ *       {"name": "key", type="RecordKeySchema"},
+ *       {"name": "offset", type="long"},
+ *       {"name": "timestamp", type="long"},
+ *       {"name": "headers", type="map"},
+ *       {"name": "value", type="RecordValueSchema"},
+ *    ]
+ * }
+ */
+public abstract class SinkSchemaBuilder {
+
+    private final Logger logger = LoggerFactory.getLogger(SinkSchemaBuilder.class);
+
+    private final Collection<OutputField> fields;
+
+    private final AvroData avroData;
+
+    private final boolean envelopeEnabled;
+
+    public SinkSchemaBuilder(final Collection<OutputField> fields,
+                      final AvroData avroData,
+                      final boolean envelopeEnabled) {
+        this.fields = fields;
+        this.avroData = avroData;
+        this.envelopeEnabled = envelopeEnabled;
+    }
+
+    public SinkSchemaBuilder(final Collection<OutputField> fields,
+                                final AvroData avroData) {
+        this.fields = fields;
+        this.avroData = avroData;
+        this.envelopeEnabled = true;
+    }
+
+    protected abstract String getNamespace();
+
+    public Schema buildSchema(final SinkRecord record) {
+        Objects.requireNonNull(record, "record");
+        if (Objects.isNull(record.keySchema())) {
+            throw new DataException("Record key without schema");
+        }
+        if (Objects.isNull(record.valueSchema())) {
+            throw new DataException("Record value without schema");
+        }
```

Comment:
Why are two different exceptions thrown here? Couldn't we stick to either NPE or DataException?